### PR TITLE
[JENKINS-37324] API to format Step arguments to string

### PR DESCRIPTION
**Pull request by svanoort**
Part of [JENKINS-37324](https://issues.jenkins-ci.org/browse/JENKINS-37324) and upstream of:

https://github.com/jenkinsci/workflow-api-plugin/pull/26
https://github.com/jenkinsci/workflow-cps-plugin/pull/98

### CODE REVIEW

**Comment by svanoort** [comment]
Cut a beta as 2.10-stepstorage4-beta for downstream dependencies to consume (sigh).

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 1]
```diff
@@ -305,6 +306,27 @@ public boolean apply(StepDescriptor i) {
         return r;
     }
 
+    /** Return true if we can easily create a nice String for user display from the object */
+    static boolean isAbleToUseToStringForDisplay(@CheckForNull Object o) {
+        return (o != null && (o instanceof Comparable || o instanceof CharSequence)); // Covers our base types
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: `Comparable` seems dodgy to me, there is no contract with `Comparable` that will ensure general utility... I think `Number` or `Boolean` is really what you want for this, e.g. `(o instanceof CharSequence || o instanceof Number || o instanceof Boolean)`

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: also `instanceof` already covers the `!= null` case

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 2]
```diff
@@ -305,6 +306,27 @@ public boolean apply(StepDescriptor i) {
...
+    }
+
+    /**
+     * Converts user-supplied step arguments to a string for eventual UI use -- override to handle more than a single trivial argument.
+     * Complements {@link #getDisplayName()} in cases where the step type is less meaningful than its arguments (scripts, for example).
+     * <em>Note: this offers a raw value and does not perform escaping on its own.</em>
+     * @param namedArgs List of parameter name-value pairs supplied to the step to instantiate it, as from {@link #uninstantiate(Step)}
+     * @return Formatted string, before escaping, or null if can't be converted easily to a String.
+     */
+    @Nullable
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: `@Nullable` is the default for this package. Given that `null` is an expected outcome it may be better to either use `@CheckForNull` or omit the annotation.

Most people get confused by `@Nullable` saying that 

> The annotated element could be `null` under some circumstances. In general, this means developers will have to read the documentation to determine when a `null` value is acceptable and whether it is necessary to check for a `null` value.

The documentation you provide does not define the full set of criteria when `null` will be returned, e.g. doesn't cover the case where the single argument value is `null` nor the case where there are multiple arguments nor the case where the supplied map is `null`... and then overridden implementations may have entirely different criteria under which the returned value will be `null`... and I cannot read all implementations.

In other words (and this is why I am negging on this specific use of an annotation) this method is *both* an API and an SPI, thus you either should document strictly the cases for returning `null` so that implementers have a strong contract to follow (i.e. make the SPI strict enough that the documentation is complete in the method itself) or switch to an API/SPI split (which is not the existing pattern in this class and will likely cause confusion)... or maybe switch to `@CheckForNull`

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 1 above)
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Yes, I was just on my way back to suggest `Number` and `Boolean` as well.

File: pom.xml [hunk 3]
```diff
@@ -33,7 +33,8 @@
     </parent>
     <groupId>org.jenkins-ci.plugins.workflow</groupId>
     <artifactId>workflow-step-api</artifactId>
-    <version>2.10-SNAPSHOT</version>
+    <!-- Remove before merge, used to identify specific version for linked PRs, will cut as a beta -->
```
Comment (jglick) [inline pom.xml]: Generally prefer that you `deploy` timestamped snapshots instead, so that the list of release versions in the repository is not permanently cluttered by experimental code (unless you are really intending to offer builds to users on the experimental UC).

Also BTW please file PRs from forks rather than from the origin repository.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Or primitive types more generally (there are various utilities say in Commons Lang for that).

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 4]
```diff
@@ -305,6 +306,27 @@ public boolean apply(StepDescriptor i) {
         return r;
     }
 
+    /** Return true if we can easily create a nice String for user display from the object */
+    static boolean isAbleToUseToStringForDisplay(@CheckForNull Object o) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Prefer that arguments be `@Nonnull` unless there is a compelling reason why it would be cumbersome for callers to do their own null handling, or there is some meaningful need to pass nulls (which apparently here there is not).

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: And why `CharSequence` rather than simply `String`?

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 5]
```diff
@@ -305,6 +306,27 @@ public boolean apply(StepDescriptor i) {
...
+        return (o != null && (o instanceof Comparable || o instanceof CharSequence)); // Covers our base types
+    }
+
+    /**
+     * Converts user-supplied step arguments to a string for eventual UI use -- override to handle more than a single trivial argument.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Split into summary sentence ending in `.` and additional descriptive sentences.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 6]
```diff
@@ -305,6 +306,27 @@ public boolean apply(StepDescriptor i) {
...
+     * Complements {@link #getDisplayName()} in cases where the step type is less meaningful than its arguments (scripts, for example).
+     * <em>Note: this offers a raw value and does not perform escaping on its own.</em>
+     * @param namedArgs List of parameter name-value pairs supplied to the step to instantiate it, as from {@link #uninstantiate(Step)}
+     * @return Formatted string, before escaping, or null if can't be converted easily to a String.
+     */
+    @Nullable
+    public String argumentsToString(@CheckForNull Map<String, Object> namedArgs) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: `namedArgs` should be `@Nonnull`. The suggested origin method would not return null.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java [hunk 7]
```diff
@@ -305,6 +306,27 @@ public boolean apply(StepDescriptor i) {
...
+    /** Return true if we can easily create a nice String for user display from the object */
+    static boolean isAbleToUseToStringForDisplay(@CheckForNull Object o) {
+        return (o != null && (o instanceof Comparable || o instanceof CharSequence)); // Covers our base types
+    }
+
+    /**
+     * Converts user-supplied step arguments to a string for eventual UI use -- override to handle more than a single trivial argument.
+     * Complements {@link #getDisplayName()} in cases where the step type is less meaningful than its arguments (scripts, for example).
+     * <em>Note: this offers a raw value and does not perform escaping on its own.</em>
+     * @param namedArgs List of parameter name-value pairs supplied to the step to instantiate it, as from {@link #uninstantiate(Step)}
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Do you mean `defineArguments` (deprecated)? `uninstantiate` returns an `UninstantiatedDescribable`.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Recommend using `@CheckForNull`. There is a small set of special circumstances where `@Nullable` is appropriate. This is not one of them.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 6 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: I think the method name here is misleading. Probably you would want, say,

```groovy
sh script: 'git rev-parse HEAD', returnStdout: true
```

to return `git rev-parse HEAD`, right? In which case `ShellStep.DescriptorImpl` etc. should be overriding this method to

```java
return namedArgs.get("script");
```

But that is not a representation of the argument**s** in total (which a reader of Javadoc might reasonably expect to look something like `Map.toString`); only a summary of what a user looking at a build log might be most interested in, for purposes of identifying this `Step` among others.

So I think the method name should more clearly reflect that this is a kind of instance-sensitive alternative to `StepDescriptor.getDisplayName`, as the Javadoc hints at.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 7 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Or perhaps better to refer to `newInstance`, which is still used and takes a `Map<String, Object>`.

**Comment by jglick** [comment]
You can also consider handling `isMetaStep()` automatically. Particular implementations such as `CoreStep.DescriptorImpl` could do something like

```java
public String theNewMethod(Map<String, Object> args) {
    Object delegate = args.get("delegate");
    if (delegate instanceof Map) {
        return super.theNewMethod((Map<String, Object>) delegate);
    } else if (delegate instanceof UninstantiatedDescribable) {
        return super.theNewMethod(((UninstantiatedDescribable) delegate).getArguments());
    } else {
        return null;
    }
}
```

but it is awkward for every metastep to need to copy such code.

Needed for example to ensure that

```groovy
archiveArtifacts 'target/*.war'
```

returns something sensible by default. From a user PoV there is no reason why

```groovy
stash 'target/*.war'
```

should work but a metastep should not.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 1 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: @stephenc There is no way to determine whether a toString method actually returns something useful to the user (even checking for overrides is no guarantee).  Comparable and CharSequence are not *perfect* but they cover 90% of cases, and a less-than-useful result here only results in a bit of visual ugliness.  In such cases we can always have the StepDescriptor provide its own implementation.     Number, Boolean, and most classes with a useful toString tend to implement Comparable.

  Initially I explored what you are suggesting, and them realized how many cases would be rejected if you only whitelisted a handful of types.  You also forgot Enums, for example, and some other Jenkins objects potentially.  I can guarantee there are others.  

CharSequence allows for other serializable character types besides Strings.  

Instanceof *does* cover the null case, but putting that check up front short-circuits a complex chain of conditionals.

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: Well how often do you expect a null single value parameter? Would seem very rare that there is a step taking a single parameter and nobody provides a value for it. So I suspect the 95% case will be non-null and you hit the chain anyway... smells like premature optimisation.

As to useful toStrings... we want implenters to actually override the method in the 60% of cases. Much easier to implement a strict whitelist and extend case by case when there is evidence of general safety... you are exposing the general first and when you lock down then plugin authors have a "breaking" change.

Make your SPI strict to start and only open up as you need to. So I recommend locking this stuff *down* as this is an SPI

Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: @stephenc Actually we want implementers to override *this* method very rarely.  

What they should be overriding is argumentsToString and doing their own explicit logic -- in which case they won't be invoking it at all.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java (hunk 4 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/StepDescriptor.java]: I don't want to force callers to do their own checking.

**Comment by svanoort** [comment]
@reviewbybees

**Comment by svanoort** [comment]
@jglick I disagree that method makes sense as a top-level API, especially since all the MetaSteps will have to repeat the same logic (and it's unlikely that we'll see additional variants).  This is more of an implementation detail than anything else, so it belongs with the implementation itself.   Furthermore, the DSL class in workflow-cps (see invokeDescribable) has implementation logic specific to meta-step.

Your logic is heard and your feedback appreciated, but for these reasons I will be working on hardening the implementation against future changes rather than adding a net-new API here.

**Comment by jglick** [comment]
> makes sense as a top-level API

Perhaps I was unclear. I was proposing only that the default implementation of `argumentsToString` handle metasteps. I do not think that requires any additional API.

> all the MetaSteps will have to repeat the same logic

Not unless they need to override that default implementation.

**Comment by svanoort** [comment]
> CoreStep.DescriptorImpl

Correct, the expectation is that the default argumentsToString implementation will only handle the most trivial cases.  Anything fancier is up to the StepDescriptors to deal with.
